File: schema/javascript/easemlschema/test/node/dataset_test.js
'use strict'

import sch from '../../src/schema'
import ds from '../../src/dataset'
import fsOpener from '../../src/openers/fs-opener'

import { expect } from 'chai'

import fs from 'fs'
import path from 'path'

const relTestPath = path.join(__dirname, '../../../../test-examples')
const datasetInferPositivePath = 'dataset/infer/positive'
const datasetInferNegativePath = 'dataset/infer/negative'
const datasetGenPositivePath = 'dataset/generate/positive'

describe('Dataset', function () {
  describe('#infer', function () {
    describe('#positive', function () {
      var files = fs.readdirSync(path.join(relTestPath, datasetInferPositivePath))
      for (let i = 0; i < files.length; i++) {
        let dirName = files[i]
        if (fs.lstatSync(path.join(relTestPath, datasetInferPositivePath, dirName)).isDirectory()) {
          let testName = path.join(datasetInferPositivePath, dirName)
          let exampleFile = path.join(relTestPath, datasetInferPositivePath, dirName + '.json')
          let exampleDir = path.join(relTestPath, datasetInferPositivePath, dirName)
          it(testName, testDatasetInfer(exampleFile, exampleDir, true))
        }
      }
    })
    describe('#negative', function () {
      var files = fs.readdirSync(path.join(relTestPath, datasetInferNegativePath))
      for (let i = 0; i < files.length; i++) {
        let dirName = files[i]
        if (fs.lstatSync(path.join(relTestPath, datasetInferNegativePath, dirName)).isDirectory()) {
          let testName = path.join(datasetInferNegativePath, dirName)
          let exampleDir = path.join(relTestPath, datasetInferNegativePath, dirName)
          it(testName, testDatasetInfer(null, exampleDir, false))
        }
      }
    })
  })
  describe('#infer', function () {
    describe('#positive', function () {
      var files = fs.readdirSync(path.join(relTestPath, datasetGenPositivePath))
      for (let i = 0; i < files.length; i++) {
        if (files[i].endsWith('.json')) {
          let testName = path.join(datasetGenPositivePath, files[i].slice(0, -'.json'.length))
          let exampleFile = path.join(relTestPath, datasetGenPositivePath, files[i])
          it(testName, testDatasetGenerate(exampleFile))
        }
      }
    })
  })
})

function testDatasetInfer (exampleFile, exampleDir, positiveExample) {
  return function () {
    let caughtError = null
    let srcSchema = null

    try {
      // Load the dataset.
      let dataset = ds.load(exampleDir, fsOpener.defaultOpener, true)

      // Infer the schema.
      srcSchema = dataset.inferSchema()
    } catch (error) {
      if (positiveExample) {
        throw error
      } else {
        caughtError = error
        if ((caughtError instanceof ds.DatasetException) === false) {
          throw error
        }
      }
    }

    if (positiveExample) {
      // There should have been no errors.
      expect(caughtError).to.be.null

      // Load the JSON file and the schema.
      let dstData = fs.readFileSync(exampleFile)
      let dst = JSON.parse(dstData)
      let dstSchema = sch.load(dst)

      // Match the loaded schema to the dataset.
      let match = dstSchema.match(srcSchema)
      expect(match).to.be.true
    } else {
      // The caught error should be a dataset exception.
      expect(caughtError).to.be.instanceOf(ds.DatasetException)
    }
  }
}

function testDatasetGenerate (exampleFile) {
  return function () {
    // Load the JSON file and the schema.
    let dstData = fs.readFileSync(exampleFile)
    let dst = JSON.parse(dstData)
    let dstSchema = sch.load(dst)

    // Generate random dataset from the schema.
    let dataset = ds.generateFromSchema('root', dstSchema, 10, 10)

    // Infer the schema and match.
    let srcSchema = dataset.inferSchema()
    let match = dstSchema.match(srcSchema, false)
    expect(match).to.be.true
  }
}

File: schema/go/easemlschema/dataset/dataset_test.go
package dataset

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"path"
	"strings"
	"testing"

	sch "github.com/ds3lab/easeml/schema/go/easemlschema/schema"
)

// Here we assume all test will be ran from the v0.1/go directory.
const relTestPath = "../../../test-examples"
const datasetInferPositivePath = "dataset/infer/positive"
const datasetInferNegativePath = "dataset/infer/negative"
const datasetGenPositivePath = "dataset/generate/positive"

func TestDatasetInfer(t *testing.T) {

	// Load positive test examples.
	files, err := ioutil.ReadDir(path.Join(relTestPath, datasetInferPositivePath))
	if err != nil {
		log.Fatal(err)
	}
	for i := range files {
		if files[i].IsDir() {
			var dirName = files[i].Name()
			var fileName = dirName + ".json"
			var testName = path.Join(datasetInferPositivePath, dirName)
			var exampleFile = path.Join(relTestPath, datasetInferPositivePath, fileName)
			var exampleDir = path.Join(relTestPath, datasetInferPositivePath, dirName)

			t.Run(testName, testDatasetInfer(exampleFile, exampleDir, true))
		}
	}

	// Load negative test examples.
	files, err = ioutil.ReadDir(path.Join(relTestPath, datasetInferNegativePath))
	if err != nil {
		log.Fatal(err)
	}
	for i := range files {
		if files[i].IsDir() {
			var dirName = files[i].Name()
			var testName = path.Join(datasetInferNegativePath, dirName)
			var exampleDir = path.Join(relTestPath, datasetInferNegativePath, dirName)

			t.Run(testName, testDatasetInfer("", exampleDir, false))
		}
	}

}

func testDatasetInfer(exampleFile string, exampleDir string, positiveExample bool) func(*testing.T) {
	return func(t *testing.T) {

		// Load the dataset.
		dataset, err := Load(exampleDir, true, DefaultOpener{})
		if err != nil {
			if positiveExample {
				t.Error("Dataset load failed.")
			}
			return
		}

		// Infer the schema.
		srcSchema, err := dataset.InferSchema()

		if positiveExample {
			if err != nil {
				dsErr, ok := err.(Error)
				if ok == false {
					t.Error("Resulting error should be a dataser.Error instance.")
				}
				msg := dsErr.Error()
				pth := dsErr.Path()

				t.Errorf("Found error '%s' on path '%s'.", msg, pth)
				return
			}

			// Load the file.
			dstData, err := ioutil.ReadFile(exampleFile)
			if err != nil {
				panic(err)
			}

			// Unmarshal JSON into a map.
			var dst map[string]interface{}
			err = json.Unmarshal(dstData, &dst)
			if err != nil {
				panic(err)
			}

			// Load the schema.
			dstSchema, err := sch.Load(dst)
			if err != nil {
				panic(err)
			}

			// Make sure the inferred dataset matches the schema.
			match, _ := dstSchema.Match(srcSchema, false)

			if match == false {
				t.Error("Schema match failed.")
				return
			}

		} else {

			if err == nil {
				t.Error("Expected an error but got none.")
			} else {
				_, ok := err.(Error)
				if ok == false {
					t.Error("Resulting error should be a dataser.Error instance.")
				}
			}
		}
	}
}

func TestDatasetGenerate(t *testing.T) {

	// Load positive test examples.
	files, err := ioutil.ReadDir(path.Join(relTestPath, datasetGenPositivePath))
	if err != nil {
		log.Fatal(err)
	}
	for i := range files {
		var fileName = files[i].Name()
		var testName = path.Join(datasetGenPositivePath, strings.TrimSuffix(fileName, ".json"))
		var exampleFile = path.Join(relTestPath, datasetGenPositivePath, fileName)

		t.Run(testName, testDatasetGenerate(exampleFile))
	}

}

func testDatasetGenerate(exampleFile string) func(*testing.T) {
	return func(t *testing.T) {

		// Load the file.
		dstData, err := ioutil.ReadFile(exampleFile)
		if err != nil {
			panic(err)
		}

		// Unmarshal JSON into a map.
		var dst map[string]interface{}
		err = json.Unmarshal(dstData, &dst)
		if err != nil {
			panic(err)
		}

		// Load the schema.
		dstSchema, err := sch.Load(dst)
		if err != nil {
			panic(err)
		}

		// Generate sample names.
		sampleNames := make([]string, 10)
		for i := range sampleNames {
			sampleNames[i] = RandomString(10, "")
		}

		// Generate random dataset from the schema.
		dataset, err := GenerateFromSchema("root", dstSchema, sampleNames, 10)
		if err != nil {
			t.Error("Schema generation error. " + err.Error())
			return
		}

		// Infer the schema.
		srcSchema, err := dataset.InferSchema()
		if err != nil {
			t.Error("Schema inference error. " + err.Error())
			return
		}

		// Make sure the inferred dataset matches the schema.
		match, _ := dstSchema.Match(srcSchema, false)

		if match == false {
			t.Error("Schema match failed.")
			return
		}
	}
}
